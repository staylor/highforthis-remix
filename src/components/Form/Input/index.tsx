import type { ChangeEvent, InputHTMLAttributes } from 'react';
import cn from 'classnames';

import { inputBase } from '@/components/Form/styles';

export interface InputProps extends Omit<InputHTMLAttributes<HTMLInputElement>, 'onChange'> {
  onChange?: (value: string) => void;
  inputType?: 'url' | 'email' | 'number' | 'password';
}

export default function Input({ inputType, value, onChange, className, ...props }: InputProps) {
  const inputOnChange = (e: ChangeEvent<HTMLInputElement>) => {
    if (!onChange) {
      return;
    }

    const newValue = e.target.value || '';
    onChange(newValue);
  };

  return (
    <input
      type={inputType || 'text'}
      {...props}
      onChange={onChange ? inputOnChange : undefined}
      className={cn(inputBase, 'w-full', className)}
      defaultValue={value}
    />
  );
}
