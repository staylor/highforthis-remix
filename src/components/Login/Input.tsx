import type { InputHTMLAttributes } from 'react';

const Input = (props: InputHTMLAttributes<HTMLInputElement>) => (
  <input
    {...props}
    className="mb-5 mr-1.5 mt-0.5 box-border w-full rounded border border-gray-300 bg-gray-50 p-2 text-2xl shadow-sm outline-none transition focus:border-gray-400 focus:shadow-lg"
  />
);

export default Input;
