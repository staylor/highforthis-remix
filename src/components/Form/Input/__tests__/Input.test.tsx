import { describe, expect, test, vi } from 'vitest';
import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Input from '@/components/Form/Input';

const TEXT_VALUE = 'Run for the border.';

describe('Input', () => {
  test('empty', () => {
    const { container } = render(<Input />);

    expect(container.firstChild).toMatchSnapshot();
    expect((container.firstChild as HTMLInputElement).value).toEqual('');
  });

  test('add className', () => {
    // eslint-disable-next-line tailwindcss/no-custom-classname
    const { container } = render(<Input className="foo" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  describe('onChange', () => {
    test('adding text', async () => {
      const user = userEvent.setup();
      const func = vi.fn();
      const value = TEXT_VALUE;
      render(<Input onChange={func} />);

      await user.type(screen.getByRole('textbox'), value);

      expect(func).toHaveBeenCalledTimes(value.length);
    });

    test('removing text', async () => {
      const user = userEvent.setup();
      const func = vi.fn();
      const value = TEXT_VALUE;
      render(<Input onChange={func} value={value} />);

      await user.type(screen.getByRole('textbox'), 'f');

      expect(func).toHaveBeenCalledWith(value + 'f');
    });
  });
});
