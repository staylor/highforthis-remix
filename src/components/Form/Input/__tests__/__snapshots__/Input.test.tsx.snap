// Vitest Snapshot v1

exports[`Input > add className 1`] = `
<input
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors w-full foo"
  defaultvalue=""
  type="text"
/>
`;

exports[`Input > empty 1`] = `
<input
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors w-full"
  defaultvalue=""
  type="text"
/>
`;
